muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java

```diff
@@ -1,71 +1,315 @@
-    
+
     return Response.ok().type(contentType).entity(output).build();
   }
-  
+
+  @GET
+  @Path("/vops/{IDENTIFIER}")
+  @RESTPermit(handling = Handling.INLINE)
+  public Response getVops(@PathParam("IDENTIFIER") String studentIdentifierString) {
+
+    String educationTypeMappingString = pluginSettingsController.getPluginSetting("transcriptofrecords", "educationTypeMapping");
+    EducationTypeMapping educationTypeMapping = new EducationTypeMapping();
+    if (educationTypeMappingString != null) {
+      try {
+        educationTypeMapping = new ObjectMapper().readValue(educationTypeMappingString, EducationTypeMapping.class);
+      } catch (IOException e) {
+        logger.log(Level.INFO, "Education type mapping not set");
+      }
+    }
+
+    if (!sessionController.isLoggedIn()) {
+      return Response.status(Status.FORBIDDEN).entity("Must be logged in").build();
+    }
+
+    SchoolDataIdentifier studentIdentifier = SchoolDataIdentifier.fromId(studentIdentifierString);
+
+    if (studentIdentifier == null) {
+      return Response.status(Status.NOT_FOUND).entity("Student identifier not found").build();
+    }
+
+    if (!Objects.equals(sessionController.getLoggedUser(), studentIdentifier)) {
+      return Response.status(Status.FORBIDDEN).entity("Can only look at own information").build();
+    }
+
+    User student = userController.findUserByIdentifier(studentIdentifier);
+
+    List<Subject> subjects = courseMetaController.listSubjects();
+    List<VopsRESTModel.VopsRow> rows = new ArrayList<>();
+    int numCourses = 0;
+    int numMandatoryCourses = 0;
+    
+    for (Subject subject : subjects) {
+      if (vopsController.subjectAppliesToStudent(student, subject)) {
+        List<VopsRESTModel.VopsItem> items = new ArrayList<>();
+        for (int i=1; i<MAX_COURSE_NUMBER; i++) {
+          List<Workspace> workspaces =
+              workspaceController.listWorkspacesBySubjectIdentifierAndCourseNumber(
+                  subject.getSchoolDataSource(),
+                  subject.getIdentifier(),
+                  i);
+          List<WorkspaceAssessment> workspaceAssessments = new ArrayList<>();
+          if (!workspaces.isEmpty()) {
+            SchoolDataIdentifier educationSubtypeIdentifier = null;
+            boolean workspaceUserExists = false;
+            for (Workspace workspace : workspaces) {
+              WorkspaceEntity workspaceEntity =
+                  workspaceController.findWorkspaceEntity(workspace);
+              WorkspaceUserEntity workspaceUser =
+                  workspaceUserEntityController.findWorkspaceUserByWorkspaceEntityAndUserIdentifier(
+                      workspaceEntity,
+                      studentIdentifier);
+              SchoolDataIdentifier workspaceIdentifier = new SchoolDataIdentifier(workspace.getIdentifier(), workspace.getSchoolDataSource());
+              workspaceAssessments.addAll(gradingController.listWorkspaceAssessments(workspaceIdentifier, studentIdentifier));
```

Comment:
Even if Pyramus is capable of storing multiple assessments per workspace, only the latest one should be considered the one that actually matters. At least the order of assessments retrieved via gradingController should be checked, and it probably wouldn't hurt to have a method that simply returns the most recent workspace assessment, if any.